File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.java
import java.util.*;
class Solution {
    static int[][] visited;
    static int[] dy = {-1, 0, 1, 0};
    static int[] dx = {0, 1, 0, -1};
    static boolean check;
    static char[][] temp;
    public List<Integer> solution(String[][] places) {
        List<Integer> ret = new ArrayList<>();
        for(String[] sArr : places){
            int row = sArr.length;
            int col = sArr[0].length();
            temp = new char[row][col];
            
            for(int i = 0; i < row; i++){
                temp[i] = sArr[i].toCharArray();
            }
            
            //--------
            
            // 초기화 
            check = false;
            visited = new int[row][col];
            
            for(int i = 0; i < row; i++){
                for(int j = 0; j < col; j++){
                    if(temp[i][j] == 'P') {
                        go(i, j, 2, row, col);
                    }
                    if(check) break;
                }
            }
            if(check) ret.add(0);
            else ret.add(1);
            
        }    
        return ret;
    }
    private void go(int y, int x, int cnt, int n, int m){
        if(cnt == 0) return;
        if(check) return;
        //방문처리
        visited[y][x] = 1;
        for(int i = 0; i < 4; i++){
            int ny = y + dy[i];
            int nx = x + dx[i];
            // 경계값 처리
            if(ny < 0 || nx < 0 || ny >= n || nx >= m) continue;
            if(visited[ny][nx] != 0) continue;
            // 파티션 있으면 더이상 못간다
            if(temp[ny][nx] == 'X') continue;
            // 범위 안에 사람 있으면 true
            if(temp[ny][nx] == 'P') check = true;
            go(ny, nx, cnt - 1, n, m);
        }
        visited[y][x] = 0;
    }
    
}

/*
- 대기실 5개, 5*5 크기
- 맨해튼 거리(|y1-y2| + |x1-x2|) > 2
- 파티션 있으면 허용
크기2의 dfs 탐색을 할 동안 
*/

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.py

def go(y, x, place, vis, cnt):
    dy, dx = [-1, 0, 1, 0], [0, 1, 0, -1]
    vis[y][x] = 1 
    if cnt == 2: return 0
    ret = 0
    for i in range(4):
        ny, nx = y + dy[i], x + dx[i]
        if ny < 0 or nx < 0 or ny >= 5 or nx >= 5: continue
        if place[ny][nx] == 'X' or vis[ny][nx]: continue
        if place[ny][nx] == 'P': return 1
        ret = ret or go(ny, nx, place, vis, cnt+1)
    return ret

def check(place):
    for i in range(5):
        for j in range(5):
            if place[i][j] == 'P':
                vis = [[0]*5 for _ in range(5)]
                if go(i, j, place, vis, 0): 
                    return 0
    return 1

def solution(places):
    return [check(place) for place in places]
"""
응시자들 끼리 맨해튼 거리 2 이하, 파티션으로 가로막히면 허용
거리 2 이내에 내가 도달할 수 있는 사람이 있으면 거리두기 실패
거리두기 지키면 1, 아니면 0

1. 모든 자리배치에 대해 다음을 반복
  1-1. 거리 2 이내에 내가 도달할 수있는 사람이 있는지 확인
  1-2. 있으면 0, 아니면 1 리턴
2. 배열 반환

"""